yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,412 @@
+            likelihood = likelihood + alignment_matrix.transpose(1, 2)
+            likelihood = likelihood.logsumexp(dim=-1, keepdim=True).transpose(
+                1, 2
+            )
+            pred, likelihood = self.greedy_step(log_probs, likelihood)
+            finished = finished | (pred == self.end_idx)
+            if finished.all().item():
+                break
+            # If finished decoding, pads.
+            pred = torch.where(
+                ~finished, pred, torch.tensor(self.end_idx, device=self.device)
+            )
+            predictions = torch.cat((predictions, pred), dim=-1)
+            # Updates likelihood emissions.
+            likelihood = likelihood + self._gather_at_idx(log_probs, pred)
+        return predictions, likelihood
+
+    def greedy_step(self, log_probs, likelihood):
+        # Updates current likelihood with greedy probability prediction.
+        tgt_prob = likelihood + log_probs.transpose(1, 2)
+        tgt_prob = tgt_prob.logsumexp(dim=-1)
+        tgt = torch.max(tgt_prob, dim=-1)[1]
+        return tgt.unsqueeze(-1), likelihood
+
+    def _gather_at_idx(self, prob, tgt):
+        # Collects probability of tgt index across all states in prob.
+        batch_size, src_seq_len, vocab_size = prob.shape
+        idx = tgt.view(-1, 1).expand(batch_size, src_seq_len).unsqueeze(-1)
+        output = torch.gather(prob, -1, idx).view(batch_size, 1, src_seq_len)
+        idx = idx.view(batch_size, 1, src_seq_len)
+        pad_mask = (idx != self.pad_idx).float()
+        return output * pad_mask
+
+    def _apply_mono_mask(self, alignment):
+        # Applies mask to enforce monotonic alignment.
+        align_mask = torch.ones_like(alignment[0]).triu().unsqueeze(0)
+        align_mask = (align_mask - 1) * -numpy.log(1e-7)
+        alignment = alignment + align_mask
+        alignment = alignment - alignment.logsumexp(-1, keepdim=True)
+        return alignment
+
+    def training_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
+        """Runs one step of training.
+
+        This is called by the PL Trainer.
+
+        Args:
+                batch (data.PaddedBatch)
+                batch_idx (int).
+
+        Returns:
+                torch.Tensor: loss.
+        """
+        # Forward pass produces loss by default.
+        log_probs, alignments = self(batch)
+        loss = self.loss_func(batch.target.padded, log_probs, alignments)
+        self.log(
+            "train_loss",
+            loss,
+            batch_size=len(batch),
+            on_step=False,
+            on_epoch=True,
+        )
+        return loss
+
+    def validation_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
+        predictions, likelihood = self(batch)
+        # Processes for accuracy calculation.
+        predictions = self.evaluator.finalize_predictions(
+            predictions, self.end_idx, self.pad_idx
+        )
+        val_eval_item = self.evaluator.get_eval_item(
```

Comment:
There are some updates here in master. I can fix this up later though.

Reply:
Any major ones?

Reply:
Yeah https://github.com/CUNY-CL/yoyodyne/pull/184

Reply:
Ah yeah, I updated it that in a push. Check that in current file.